jetty-util/src/main/java/org/eclipse/jetty/util/resource/ZipFsResource.java

```diff
@@ -0,0 +1,306 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2018 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util.resource;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.DirectoryStream;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystemNotFoundException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+public class ZipFsResource extends Resource implements Iterable<String>
+{
+    private static final Logger LOG = Log.getLogger(ZipFsResource.class);
+    /**
+     * For a JVM that support JEP-238 Multi-Release jars, use "runtime" behavior.
+     * On older JVMs that don't have JEP-238 support / behavior (eg: Java 1.8), this
+     * FileSystem Environment variable is not used. (it's ignored)
+     */
+    private static final String MR_MODE_RUNTIME = "runtime";
+
+    private final URI zipFsUri;
+    private final Path zipFsPath;
+    private final Path rootPath;
```

Comment:
isn't the root path always `/`???  is this the `/other/path` of `jar:file:///some/path/file.zip!/other/path`? if so, then perhaps `path`, `resourcePath` or `entryPath` would be better names?

Reply:
All FileSystems have 1 (or more) roots.
While a JAR file traditionally has 1 root, the fact that the FileSystem supports more then 1 root makes me leery to assume anything (yet).

Note: a FileSystem has Path objects.  A Path object has a reference back to its FileSystem.

The rootPath is just a good place to hold onto the Path that belongs to *this* FileSystem.

Reply:
So the rootPath for `file:///some/path/file.zip` is `jar:file:///some/path/file.zip!/` ?
If so, then the name is good!